runtime-light/stdlib/rpc/rpc-api.cpp

```diff
@@ -0,0 +1,522 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2024 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+#include "runtime-light/stdlib/rpc/rpc-api.h"
+
+#include <array>
+#include <chrono>
+#include <coroutine>
+#include <cstddef>
+#include <optional>
+#include <utility>
+
+#include "common/algorithms/find.h"
+#include "common/rpc-error-codes.h"
+#include "common/tl/constants/common.h"
+#include "runtime-light/allocator/allocator.h"
+#include "runtime-light/stdlib/rpc/rpc-context.h"
+#include "runtime-light/stdlib/rpc/rpc-extra-headers.h"
+#include "runtime-light/streams/interface.h"
+#include "runtime-light/utils/concepts.h"
+
+namespace rpc_impl_ {
+
+constexpr int32_t MAX_TIMEOUT_S = 86400;
+
+// TODO: change uint64_t to string::size_type after moving it from uint32_t to uint64_t
+constexpr uint64_t SMALL_STRING_MAX_LEN = 253;
+constexpr uint64_t MEDIUM_STRING_MAX_LEN = (static_cast<uint64_t>(1) << 24) - 1;
+[[maybe_unused]] constexpr uint64_t LARGE_STRING_MAX_LEN = (static_cast<uint64_t>(1) << 56) - 1;
+
+constexpr uint8_t LARGE_STRING_MAGIC = 0xff;
+constexpr uint8_t MEDIUM_STRING_MAGIC = 0xfe;
+
+mixed mixed_array_get_value(const mixed &arr, const string &str_key, int64_t num_key) noexcept {
+  if (!arr.is_array()) {
+    return {};
+  }
+
+  if (const auto &elem{arr.get_value(num_key)}; !elem.is_null()) {
+    return elem;
+  }
+  if (const auto &elem{arr.get_value(str_key)}; !elem.is_null()) {
+    return elem;
+  }
+  return {};
+}
+
+bool rpc_fetch_remaining_enough(size_t len) noexcept {
+  return RpcComponentContext::get().fetch_state.remaining() >= len;
+}
+
+array<mixed> make_fetch_error(string &&error_msg, int32_t error_code) {
+  array<mixed> res;
+  res.set_value(string{"__error", 7}, std::move(error_msg));
```

Comment:
why string(char *, size_t) instead of just string(char *)?

Reply:
Since it's more efficient. We don't need to call `strlen` when we use that constructor